### PR TITLE
feat: add subscription uncancel functionality

### PR DESCRIPTION
**Pull request by angihe93**
## What Does this PR Do?

Implements the complete subscription uncancel functionality, allowing customers to cancel a cancellation that was previously scheduled. This PR includes:

- **Core uncancel logic** (`uncancelSubscription` and `uncancelSubscriptionProcedureTransaction`): Reverses scheduled cancellations with proper state management
- **Billing period restoration**: Reverts billing periods from ScheduledToCancel to Active/Upcoming based on timing
- **Billing run rescheduling**: Creates new billing runs for future periods that had aborted runs
- **Security checks**: Requires valid payment methods for paid subscriptions
- **Comprehensive tests**: 62 tests covering current/future billing periods, payment method validation, and idempotency

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
## Summary by cubic
Adds the ability to uncancel a scheduled subscription cancellation. Restores status and billing periods, and reschedules billing runs so billing continues as expected.

- **New Features**
  - Core functions: uncancelSubscription and uncancelSubscriptionProcedureTransaction.
  - Restores subscription status to Active or Trialing based on trialEnd.
  - Restores billing periods from ScheduledToCancel to Active or Upcoming.
  - Reschedules billing runs: creates new runs for periods aborted by cancellation, skips terminal/scheduled runs and active/awaiting/Stripe-aborted runs, respects runBillingAtPeriodStart, only future dates, and ignores trial periods.
  - Security: paid subscriptions require an active payment method (uses backup if available); free plans can uncancel without one.
  - Idempotent: safe to call multiple times without side effects.
  - Adds uncancelSubscriptionSchema and comprehensive tests covering timing, trials, payment method checks, and idempotency.

<sup>Written for commit 254071b99b9f239a61e6889648b17e9db0059180. Summary will update automatically on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Users can uncancel scheduled subscription cancellations to restore Active or Trial status.
  * Previously scheduled billing periods are restored and billing runs rescheduled as needed, honoring trial periods and scheduling flags.
  * Uncancel requires a valid payment method for paid plans; free plans may uncancel without one.
  * New uncancel API/schema for initiating uncancel operations.

* **Tests**
  * Comprehensive tests covering uncancel flows, billing-run behaviors, validations, event handling, idempotency, and edge cases.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: platform/flowglad-next/src/subscriptions/cancelSubscription.ts [hunk 1]
```diff
@@ -634,3 +636,272 @@ export const cancelSubscriptionProcedureTransaction = async ({
...
+    if (!paymentMethod) {
+      throw new TRPCError({
+        code: 'BAD_REQUEST',
+        message:
+          'Cannot uncancel paid subscription without an active payment method. Please add a payment method first.',
+      })
+    }
+  }
+
+  // Get payment method for billing run creation
```
[the remark this answers is not in the document]
Comment (joeysabs) [reply platform/flowglad-next/src/subscriptions/cancelSubscription.ts]: Definitely nitpicky but we can get rid of this redundant call to "selectPaymentMethodById" :)

File: platform/flowglad-next/src/subscriptions/cancelSubscription.ts [hunk 2]
```diff
@@ -634,3 +636,272 @@ export const cancelSubscriptionProcedureTransaction = async ({
...
+    const existingRuns = await selectBillingRuns(
+      { billingPeriodId: billingPeriod.id },
+      transaction
+    )
+
+    // Leave Scheduled runs as-is (already valid)
+    const hasScheduledRun = existingRuns.some(
+      (run) => run.status === BillingRunStatus.Scheduled
+    )
+    if (hasScheduledRun) continue
```
Comment (joeysabs) [inline platform/flowglad-next/src/subscriptions/cancelSubscription.ts]: Okay quick question. Why are whole billing periods being passed if we only have one run in them that is in a terminal or scheduled state?
